test/extensions/filters/network/geoip/geoip_filter_test.cc

```diff
@@ -0,0 +1,257 @@
+        filter_state_(std::make_shared<StreamInfo::FilterStateImpl>(
+            StreamInfo::FilterState::LifeSpan::Connection)) {
+    ON_CALL(filter_callbacks_.connection_.stream_info_, filterState())
+        .WillByDefault(testing::ReturnRef(filter_state_));
+  }
+
+  void initializeFilter(const std::string& yaml) {
+    envoy::extensions::filters::network::geoip::v3::Geoip config;
+    TestUtility::loadFromYaml(yaml, config);
+    config_ = std::make_shared<GeoipFilterConfig>(config, "prefix.", stats_.mockScope());
+    filter_ = std::make_shared<GeoipFilter>(config_, dummy_driver_);
+    filter_->initializeReadFilterCallbacks(filter_callbacks_);
+  }
+
+  void initializeProviderFactory() {
+    Registry::InjectFactory<Geolocation::GeoipProviderFactory> registered(*dummy_factory_);
+  }
+
+  void expectStats(const uint32_t n_total = 1) {
+    EXPECT_CALL(stats_, counter("prefix.geoip.total")).Times(n_total);
+  }
+
+  NiceMock<Stats::MockStore> stats_;
+  GeoipFilterConfigSharedPtr config_;
+  std::shared_ptr<GeoipFilter> filter_;
+  std::unique_ptr<DummyGeoipProviderFactory> dummy_factory_;
+  MockDriverSharedPtr dummy_driver_;
+  NiceMock<Network::MockReadFilterCallbacks> filter_callbacks_;
+  StreamInfo::FilterStateSharedPtr filter_state_;
+  Geolocation::LookupRequest captured_rq_;
+  Geolocation::LookupGeoHeadersCallback captured_cb_;
+};
+
+TEST_F(GeoipFilterTest, SuccessfulLookupStoresFilterState) {
+  initializeProviderFactory();
+  const std::string config_yaml = R"EOF(
+    provider:
+        name: "envoy.geoip_providers.dummy"
+        typed_config:
+          "@type": type.googleapis.com/test.extensions.filters.http.geoip.DummyProvider
+)EOF";
+  initializeFilter(config_yaml);
+
+  Network::Address::InstanceConstSharedPtr remote_address =
+      Network::Utility::parseInternetAddressNoThrow("1.2.3.4");
+  filter_callbacks_.connection_.stream_info_.downstream_connection_info_provider_->setRemoteAddress(
+      remote_address);
+
+  expectStats();
+  EXPECT_CALL(*dummy_driver_, lookup(_, _))
+      .WillOnce(DoAll(SaveArg<0>(&captured_rq_), SaveArg<1>(&captured_cb_), Invoke([this]() {
+                        captured_cb_(Geolocation::LookupResult{{"x-geo-city", "TestCity"},
+                                                               {"x-geo-country", "US"}});
+                      })));
+
+  EXPECT_EQ(Network::FilterStatus::Continue, filter_->onNewConnection());
+  EXPECT_EQ("1.2.3.4:0", captured_rq_.remoteAddress()->asString());
+
+  // Verify GeoipInfo was stored in filter state.
+  ASSERT_TRUE(filter_state_->hasData<GeoipInfo>(std::string(DefaultGeoipFilterStateKey)));
+  const auto* geoip_info =
+      filter_state_->getDataReadOnly<GeoipInfo>(std::string(DefaultGeoipFilterStateKey));
+  ASSERT_NE(nullptr, geoip_info);
+  EXPECT_EQ("TestCity", geoip_info->getGeoField("x-geo-city").value());
+  EXPECT_EQ("US", geoip_info->getGeoField("x-geo-country").value());
```

Comment:
Similarly, would suggest using a custom matcher for this so you could do
```
EXPECT_THAT(filter_state_, HasGeoField("x-geo-city", "TestCity"));
EXPECT_THAT(filter_state_, HasGeoField("x-geo-country", "US"));
```
(Which would be used in several places and save on doing the assert-not-nulls repeatedly.)

These should be easy enough to implement with a MATCHER_P2 macro, no need for a whole custom matcher class.
(Maybe a MATCHER_P3 that also takes the filter state key, and a MATCHER_P2 that calls that one with DefaultGeoipFilterStateKey, so that it also plays well with the custom key case.)